packages/synapse-react-client/src/lib/utils/hooks/useLogin.ts

```diff
@@ -98,28 +101,80 @@ export default function useLogin(
     }
   }
 
-  const submitUsernameAndPassword: UseLoginReturn['submitUsernameAndPassword'] =
-    async (username, password) => {
-      setErrorMessage(undefined)
-      try {
-        const authenticationReceipt = localStorage.getItem(
-          AUTHENTICATION_RECEIPT_LOCALSTORAGE_KEY,
-        )
-        const data = await SynapseClient.login(
-          username,
-          password,
-          authenticationReceipt,
-        )
-        if ('errorCode' in data) {
+  const {
+    mutateAsync: mutateLoginWithUsernameAndPassword,
+    isLoading: isLoadingLoginWithUsernameAndPassword,
+  } = useMutation<
+    LoginResponse | TwoFactorAuthErrorResponse,
+    SynapseClientError,
+    { username: string; password: string; authenticationReceipt: string | null }
+  >(
+    ({ username, password, authenticationReceipt }) =>
+      SynapseClient.login(username, password, authenticationReceipt),
+    {
+      onError: error => {
+        setErrorMessage(error.reason)
+      },
+      onSuccess: async loginResponse => {
+        if ('errorCode' in loginResponse) {
           setStep('VERIFICATION_CODE')
-          setTwoFaToken(data.twoFaToken)
-          setUserId(data.userId)
+          setTwoFaToken(loginResponse.twoFaToken)
+          setUserId(loginResponse.userId)
         } else {
-          await finishLogin(data)
+          await finishLogin(loginResponse)
         }
-      } catch (err) {
-        setErrorMessage(err.reason)
-      }
+      },
+    },
+  )
+
+  const {
+    mutateAsync: mutateLoginWith2FACode,
+    isLoading: isLoadingLoginWith2FACode,
+  } = useMutation<LoginResponse, SynapseClientError, TwoFactorAuthLoginRequest>(
+    SynapseClient.loginWith2fa,
+    {
+      onError: e => {
+        setErrorMessage(e.reason)
```

Comment:
This error handling feels unstable (looking at the error reason text to determine how to handle instead of checking an associated error code, and the cleanup via regex).

Reply:
https://sagebionetworks.jira.com/browse/PLFM-7709